### PR TITLE
Auto link URLs in threads and replies

### PR DESCRIPTION
**Pull request by pushpak1300**
https://github.com/laravelio/laravel.io/issues/529

### CODE REVIEW

**Comment by pushpak1300** [comment]
@driesvints Let me know if any changes are there.

File: resources/helpers.php [hunk 1]
```diff
@@ -31,3 +31,17 @@ function route_to_reply_able($replyAble): string
...
+     */
+    function linkyfy(string $markdown): string
+    {
+        $lf=new LinkFinder([
+                "attrs" => ["class" => "external-link", "target" => "_blank", "rel" => "nofollow"],
+                "mailto_attrs" => ["class" => "external-email"]
+                ]);
+        return $lf->processHtml($markdown);
+    }
+}
```
Comment (joedixon) [inline resources/helpers.php:47]:
```suggestion
if (! function_exists('replace_links')) {
    /**
     * Convert Standalone Urls to HTML.
     */
    function replace_links(string $markdown): string
    {
        return (new LinkFinder([
            "attrs" => ["target" => "_blank", "rel" => "nofollow"],
        ]))->processHtml($markdown);
    }
}
```

File: resources/views/forum/threads/show.blade.php [hunk 2]
```diff
@@ -41,7 +41,7 @@ class="text-green-darker mr-2">
                             class="forum-content" 
                             x-data="{}" 
                             x-init="function () { highlightCode($el); }"
-                            x-html="{{ json_encode(md_to_html($thread->body())) }}"
+                            x-html="{{ json_encode(linkyfy(md_to_html($thread->body()))) }}"
```
Comment (joedixon) [inline resources/views/forum/threads/show.blade.php:44]:
```suggestion
                            x-html="{{ json_encode(replace_links(md_to_html($thread->body()))) }}"
```

File: resources/views/forum/threads/show.blade.php [hunk 3]
```diff
@@ -118,7 +118,7 @@ class="forum-content"
                                     class="forum-content" 
                                     x-data="{}" 
                                     x-init="function () { highlightCode($el); }"
-                                    x-html="{{ json_encode(md_to_html($reply->body())) }}"
+                                    x-html="{{ json_encode(linkyfy(md_to_html($reply->body()))) }}"
```
Comment (joedixon) [inline resources/views/forum/threads/show.blade.php:121]:
```suggestion
                                    x-html="{{ json_encode(replace_links(md_to_html($reply->body()))) }}"
```

**Comment by pushpak1300** [comment]
@joedixon Thanks For the Review!
I'm not very familiar with the TDD, can you help me?
Can you tell me how should I structure the tests?

**Comment by driesvints** [comment]
Can you maybe update your local composer version and re attempt to install the dependency? The composer.lock file indicates that you're using an outdated version.

**Comment by pushpak1300** [comment]
@driesvints I'm using the latest version of the composer. I tried reinstalling the composer dependencies again but the same composer.lock file is getting generated.
what should I do?
Screenshot from 2020-07-22 00-53-58

**Comment by driesvints** [comment]
hmm odd. @joedixon what is your version?

**Comment by joedixon** [comment]
> hmm odd. @joedixon what is your version?

A bit behind on 1.10.5

> Can you maybe update your local composer version and re attempt to install the dependency? The composer.lock file indicates that you're using an outdated version.

What do you see that indicates an outdated version?

**Comment by driesvints** [comment]
Screenshot 2020-07-21 at 21 43 44
Screenshot 2020-07-21 at 21 43 53

**Comment by driesvints** [comment]
In general I think it's good that those funding links aren't in there.

**Comment by joedixon** [comment]
> I'm not very familiar with the TDD, can you help me?
> Can you tell me how should I structure the tests?

Hey @pushpak1300 - I would probably add a test to `tests/Feature/ForumTest.php` which uses a factory to create a thread containing a link and then use the `see` assertion to confirm it is rendered correctly when you visit the page.

**Comment by pushpak1300** [comment]
> > I'm not very familiar with the TDD, can you help me?
> > Can you tell me how should I structure the tests?
> 
> Hey @pushpak1300 - I would probably add a test to `tests/Feature/ForumTest.php` which uses a factory to create a thread containing a link and then use the `see` assertion to confirm it is rendered correctly when you visit the page.

Thanks, @joedixon! I will do that!

**Comment by joedixon** [comment]
> In general I think it's good that those funding links aren't in there.

This PR removes them though?

**Comment by driesvints** [comment]
@joedixon yeah. Wondering how they got in in the first place.

**Comment by pushpak1300** [comment]
@joedixon I have added test.
Let me know if any changes are there.

**Comment by joedixon** [comment]
Hey @pushpak1300, tests look good, I just pushed a commit to tidy them a little.

I'm happy to get this in now. Anything you want to add @driesvints?

**Comment by pushpak1300** [comment]
Thanks @joedixon !
I really enjoyed while writing the test for the first time! :smile:

**Comment by driesvints** [comment]
Thanks @pushpak1300!

**Pull request merged**